File: TouchNothing/Utilities/NicknameValidator.swift
import Foundation

enum NicknameValidationError {
    case empty
    case invalidLength
    case invalidCharacters
}

enum NicknameValidator {
    private static let allowedCharacters = CharacterSet.alphanumerics.union(CharacterSet(charactersIn: "_"))

    static func validate(_ nickname: String) -> NicknameValidationError? {
        let trimmed = nickname.trimmingCharacters(in: .whitespacesAndNewlines)
        if trimmed.isEmpty {
            return .empty
        }
        if trimmed.count < AppConstants.minNicknameLength || trimmed.count > AppConstants.maxNicknameLength {
            return .invalidLength
        }
        if trimmed.unicodeScalars.contains(where: { !allowedCharacters.contains($0) }) {
            return .invalidCharacters
        }
        return nil
    }

    static func normalized(_ nickname: String) -> String {
        nickname.trimmingCharacters(in: .whitespacesAndNewlines)
    }
}

File: app/src/main/java/com/touchnothing/app/util/NicknameValidator.kt
package com.touchnothing.app.util

enum class NicknameValidationError {
    EMPTY,
    INVALID_LENGTH,
    INVALID_CHARACTERS,
}

object NicknameValidator {
    private val allowedCharacters = Regex("^[a-zA-Z0-9_]+$")

    fun validate(nickname: String): NicknameValidationError? {
        val trimmed = normalized(nickname)
        if (trimmed.isEmpty()) return NicknameValidationError.EMPTY
        if (trimmed.length < AppConstants.MIN_NICKNAME_LENGTH ||
            trimmed.length > AppConstants.MAX_NICKNAME_LENGTH
        ) {
            return NicknameValidationError.INVALID_LENGTH
        }
        if (!allowedCharacters.matches(trimmed)) {
            return NicknameValidationError.INVALID_CHARACTERS
        }
        return null
    }

    fun normalized(nickname: String): String = nickname.trim()
}
